storage/aptosdb/src/pruner/state_store/mod.rs

```diff
@@ -121,96 +125,176 @@ where
-            // progress.
-            self.record_progress_impl(new_min_readable_version, is_end_of_target_version);
-            Ok(new_min_readable_version)
+    fn prune_single_shard(
+        &self,
+        shard_id: u8,
+        target_version: Version,
+        batch_size: usize,
+    ) -> Result<()> {
+        let _timer = OTHER_TIMERS_SECONDS
+            .with_label_values(&["state_merkle_pruner___prune_single_shard"])
+            .start_timer();
+        let shard_min_readable_version = self.get_shard_progress(shard_id);
+        if shard_min_readable_version != target_version {
+            assert_lt!(shard_min_readable_version, target_version);
+            self.update_shard_progress(shard_id, target_version);
+            let db_shard = self.state_merkle_db.db_shard(shard_id);
+            let batch = SchemaBatch::new();
+            self.prune_state_merkle_shard(
+                &db_shard,
+                shard_min_readable_version,
+                target_version,
+                batch_size,
+                &batch,
+            )?;
+            batch.put::<DbMetadataSchema>(
+                &S::tag(Some(shard_id)),
+                &DbMetadataValue::Version(target_version),
+            )?;
+            db_shard.write_schemas(batch)?;
         }
+
+        Ok(())
     }
 
-    fn record_progress_impl(&self, min_readable_version: Version, is_fully_pruned: bool) {
-        *self.progress.lock() = (min_readable_version, is_fully_pruned);
-        PRUNER_VERSIONS
-            .with_label_values(&[S::name(), "min_readable"])
-            .set(min_readable_version as i64);
+    fn prune_shards(&self, target_version: Version, batch_size: usize) -> Result<()> {
+        let num_shards = self.state_merkle_db.num_shards();
+        POOL.scope(|s| {
+            for shard_id in 0..num_shards {
+                s.spawn(move |_| {
+                    self.prune_single_shard(shard_id, target_version, batch_size)
+                        .unwrap_or_else(|_| {
+                            panic!("Failed to prune state merkle shard {shard_id}.")
+                        });
+                });
+            }
+        });
+
+        Ok(())
+    }
+
+    fn finish_pending_pruning(&self, min_readable_version: Version) -> Result<()> {
+        self.prune_shards(min_readable_version, usize::max_value())
+    }
+
+    fn get_shard_progress(&self, shard_id: u8) -> Version {
+        self.shard_progresses.lock()[shard_id as usize]
+    }
+
+    fn update_shard_progress(&self, shard_id: u8, progress: Version) {
+        self.shard_progresses.lock()[shard_id as usize] = progress;
+    }
+
+    // If the existing schema batch is not none, this function only adds items need to be
+    // deleted to the schema batch and the caller is responsible for committing the schema batches
+    // to the DB.
+    fn prune_state_merkle_shard(
+        &self,
+        db: &DB,
```

Comment:
nit: could you name this DB? too many DBs, sometime, hard to know which DB this refers to

Reply:
Done